### ISSUE TITLE
Model Skill Reporting Workflow Script

### ISSUE THREAD

**Issue by m10f** [GaloisInc/ASKE-E#76]
- [ ]  Load Data:

```
data = loadCSV("data-v1")     -- load data
times = data.time             -- get `time` column of data
```

- [ ]  Load Models:

```
model1 = loadESL("model1")
model2 = loadESL("model2")
model3 = loadESL("model3")
model4 = loadESL("model4")
```

- [ ] Build model distributions:

```
model1PDFs = pdfFromSamples(sample(model1.I at times, 100)) -- is this just `histogram?`
model2PDFs = pdfFromSamples(sample(model2.I at times, 100))
model3PDFs = pdfFromSamples(sample(model3.I at times, 100))
model4PDFs = pdfFromSamples(sample(model4.I at times, 100))
```

- [ ] Assess model skill using totally builtin WIS (option 1)

```
-- how do we synchronize the data with the pdfs?
scores = WIS([mode1PDFs, model2PDFs, model3PDFs, model4PDFs], data, [0.1, 0.2, 0.3])
```

- [ ] Compute WIS (option 2)

```
-- at some point we need to strip time off the PDFs
-- are comprehensions the right thing?
function intervalScore(pdf, observed, alpha) {
  let low = quantile(pdf, alpha/2)
  let high = quantile(pdf, 1 - alpha/2)

  let lowerScore =
    if observed < low
      then (2/alpha) * (low - observed)
      else 0

  let upperScore =
    if high < observed
      then (2/alpha) * (high - observed)
      else 0

  return (high - low) + upperScore + lowerScore
}

function WISSingle(pdfs, dataPoints, alphas) {
  let intScores =
    [ intervalScore(pdf, d, alpha) * (alpha / 2)
        for pdf in model1PDFs 
        for d in data
        if time(pdf) == d.time
        for alpha in alpha      ]
  
  let wholeModelScore = abs(y - m) / 2

  return 1/(length(alphas)+(1/2)) * wholeModelScore  + sum(intScores)
}

function WIS(pdfss, dataPoints, alphas) {
  return [WISSingle(p, dataPoints, alphas) for p in pdfss]
}


scores = WIS([model1PDFs, model2PDFs, model3PDFs, model4PDFs], data, [0.1, 0.2, 0.3])
```

- [ ] Display scores in a table

```
-- ideally, would be nice to sort these
table(["model", "score"], [["model1", "model2", "model3", "model4"], scores])
```

- [ ] Show distribution of model predictions for 7 time units in the future

```
histogram(sample(model1 at max(data.time) + 7, 100))
histogram(sample(model2 at max(data.time) + 7, 100))
histogram(sample(model3 at max(data.time) + 7, 100))
histogram(sample(model4 at max(data.time) + 7, 100))
```

- [ ] Plot model means(?) along with points on same chart

```
-- What does this look like?
```

- [ ] Copy the notebook tab, change `data` and rerun all computations - possibly tweaking one or more sets of model params

**Comment by RyanGlScott** [comment]
As much as it pains me to say, I think we should just hardcode a `wis` function (i.e., option 1).

Like you indicated in the comments, I'm unclear if there's a difference between `histogram` and what you `pdfFromSamples` (besides the fact that `pdfFromSamples` doesn't provide a number of bins).

**Comment by abakst** [comment]
Somewhat of a meta-point, but from the point of view of computing WIS, does it make sense to build a pdf over all of the samples _over time_? I would think we'd be doing comparisons at moments in time. Or is this somehow building a family of PDFs indexed by time?

**Comment by m10f** [comment]
`pdfFromSamples(sample(model1.I at times, 100))` makes an array of _timed_ PDFs with the same times as those appearing in the data - `pdfFromSamples` in this case is lifted through the top level array and the `VTimed` entries to get to the sample arrays

**Comment by m10f** [comment]
of course that requires there to be some synchronization between the data and the time intervals of the PDF - i haven't figured how this should work yet

**Comment by abakst** [comment]
After studying the metric, I realize we don't actually need `pdf`s, we can just compute quantiles from sample data.

**Comment by m10f** [comment]
I want to plot the data and some depiction of the the behavior of the model distributions on the same axes so that people can see (visually) how well each model fits and that the one with the highest score is indeed the best intuitively.

It doesn't have to be the means, and it might make sense to do a more complicated plot for each model.

**Comment by ewdavis** [comment]
> As much as it pains me to say, I think we should just hardcode a `wis` function (i.e., option 1) for now.
> 
> Like you indicated in the comments, I'm unclear if there's a difference between `histogram` and what you `pdfFromSamples` (besides the fact that `pdfFromSamples` doesn't provide a number of bins).

What part of "hard coding" WIS pains you?  It seems like the right way to expose this to me.  Our target audience isn't technical in any real sense.  Building libraries of tools they can leverage makes perfect sense to me.

Anything that looks push-button to Joshua here is likely a win IMO.

**Comment by RyanGlScott** [comment]
Don't read too deeply into that comment—I was just (partly) thinking aloud. If I had more time to significantly redesign Exposure, one thing I'd think about is allowing user-definable functions so that we could express the functionality of `wis()` in terms of more basic primitives. Of course, I don't have that kind of time before the demo, so whatever the quickest path to demonstrating the functionality is more than sufficient.

**Comment by ewdavis** [comment]
Makes sense to me!  I'm also partly just trying to incorporate this workflow and learn how to be more responsive in this fashion.

OOC, how much of a write up do we have so far on the real and aspirational parts of Exposure?

**Comment by m10f** [comment]
I think what we have is a very shaky research prototype to demonstrate the kinds of things we think demonstrate the usefulness of this general idea in a way that (hopefully) suggest that we could do something really cool with more funding.  I think it will be live demoable but not friendly enough for a scientist to actually use it, as there might be some ergonomic issues/weird behavior/bugs.

If we get ASKE-E 2 I think we'll want to figure out what we learned from this process and build a nice well-designed language

I can work on getting some documentation, but for now these scripts are fairly complete in terms of the things we can do with the language.

**Comment by abakst** [comment]
Here's the current status of this script as I understand it. I think primarily we would want to change the final plot:
[Skill Reporting (1).pdf](https://github.com/GaloisInc/ASKE-E/files/6940045/Skill.Reporting.1.pdf)

**Comment by abakst** [comment]
Crossed out pdfs/in-language WIS as tasks, since we don't need to do them :)
